src/parse/mod.rs

```diff
@@ -167,26 +167,22 @@ impl<I: Iterator<Item = Lexeme>> Iterator for Parser<I> {
     }
 }
 
+// this is just a guesstimate, it should probably be configurable
+#[cfg(debug_assertions)]
+pub(crate) const MAX_RECURSION_DEPTH: usize = 50;
+#[cfg(not(debug_assertions))]
+pub(crate) const MAX_RECURSION_DEPTH: usize = 200;
+
 impl<I: Iterator<Item = Lexeme>> Parser<I> {
     // make sure we don't crash on highly nested expressions
     // or rather, crash in a controlled way
-    fn recursion_check(&self) -> RecursionGuard {
-        // this is just a guesstimate, it should probably be configurable
-        #[cfg(debug_assertions)]
-        const MAX_DEPTH: usize = 50;
-        #[cfg(not(debug_assertions))]
-        const MAX_DEPTH: usize = 200;
-
+    fn recursion_check(&self) -> SyntaxResult<RecursionGuard> {
         let guard = self.recursion_guard.clone();
         let depth = Rc::strong_count(&guard.0);
-        if depth > MAX_DEPTH {
-            eprintln!(
-                "fatal: maximum recursion depth exceeded ({} > {})",
-                depth, MAX_DEPTH
-            );
-            std::process::exit(102);
+        if depth > MAX_RECURSION_DEPTH {
+            return Err(self.last_location.with(SyntaxError::RecursionLimit(depth)));
         }
-        guard
+        Ok(guard)
```

Comment:
There's no need for an early return here, you can use `if depth > max { Err() } else { Ok(guard) }`.